File: tests/algorithms/x/Java/data_compression/lempel_ziv_decompress.java
public class Main {
    static String sample;
    static String decompressed;

    static boolean list_contains(String[] xs, String v) {
        int i = 0;
        while (i < xs.length) {
            if ((xs[i].equals(v))) {
                return true;
            }
            i = i + 1;
        }
        return false;
    }

    static boolean is_power_of_two(int n) {
        if (n < 1) {
            return false;
        }
        int x = n;
        while (x > 1) {
            if (Math.floorMod(x, 2) != 0) {
                return false;
            }
            x = x / 2;
        }
        return true;
    }

    static String bin_string(int n) {
        if (n == 0) {
            return "0";
        }
        String res = "";
        int x_1 = n;
        while (x_1 > 0) {
            int bit = Math.floorMod(x_1, 2);
            res = _p(bit) + res;
            x_1 = x_1 / 2;
        }
        return res;
    }

    static String decompress_data(String data_bits) {
        java.util.Map<String,String> lexicon = ((java.util.Map<String,String>)(new java.util.LinkedHashMap<String, String>(java.util.Map.ofEntries(java.util.Map.entry("0", "0"), java.util.Map.entry("1", "1")))));
        String[] keys = ((String[])(new String[]{"0", "1"}));
        String result = "";
        String curr_string = "";
        int index = 2;
        int i_1 = 0;
        while (i_1 < _runeLen(data_bits)) {
            curr_string = curr_string + _substr(data_bits, i_1, i_1 + 1);
            if (!(Boolean)list_contains(((String[])(keys)), curr_string)) {
                i_1 = i_1 + 1;
                continue;
            }
            String last_match_id = ((String)(lexicon).get(curr_string));
            result = result + last_match_id;
lexicon.put(curr_string, last_match_id + "0");
            if (((Boolean)(is_power_of_two(index)))) {
                java.util.Map<String,String> new_lex = ((java.util.Map<String,String>)(new java.util.LinkedHashMap<String, String>()));
                String[] new_keys = ((String[])(new String[]{}));
                int j = 0;
                while (j < keys.length) {
                    String curr_key = keys[j];
new_lex.put("0" + curr_key, ((String)(lexicon).get(curr_key)));
                    new_keys = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(new_keys), java.util.stream.Stream.of("0" + curr_key)).toArray(String[]::new)));
                    j = j + 1;
                }
                lexicon = new_lex;
                keys = ((String[])(new_keys));
            }
            String new_key = String.valueOf(bin_string(index));
lexicon.put(new_key, last_match_id + "1");
            keys = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(keys), java.util.stream.Stream.of(new_key)).toArray(String[]::new)));
            index = index + 1;
            curr_string = "";
            i_1 = i_1 + 1;
        }
        return result;
    }
    public static void main(String[] args) {
        sample = "1011001";
        decompressed = String.valueOf(decompress_data(sample));
        System.out.println(decompressed);
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }

    static String _substr(String s, int i, int j) {
        int start = s.offsetByCodePoints(0, i);
        int end = s.offsetByCodePoints(0, j);
        return s.substring(start, end);
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/PHP/data_compression/lempel_ziv_decompress.php
<?php
ini_set('memory_limit', '-1');
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
function list_contains($xs, $v) {
  global $sample, $decompressed;
  $i = 0;
  while ($i < count($xs)) {
  if ($xs[$i] == $v) {
  return true;
}
  $i = $i + 1;
};
  return false;
}
function is_power_of_two($n) {
  global $sample, $decompressed;
  if ($n < 1) {
  return false;
}
  $x = $n;
  while ($x > 1) {
  if ($x % 2 != 0) {
  return false;
}
  $x = _intdiv($x, 2);
};
  return true;
}
function bin_string($n) {
  global $sample, $decompressed;
  if ($n == 0) {
  return '0';
}
  $res = '';
  $x = $n;
  while ($x > 0) {
  $bit = $x % 2;
  $res = _str($bit) . $res;
  $x = _intdiv($x, 2);
};
  return $res;
}
function decompress_data($data_bits) {
  global $sample, $decompressed;
  $lexicon = ['0' => '0', '1' => '1'];
  $keys = ['0', '1'];
  $result = '';
  $curr_string = '';
  $index = 2;
  $i = 0;
  while ($i < strlen($data_bits)) {
  $curr_string = $curr_string . substr($data_bits, $i, $i + 1 - $i);
  if (!list_contains($keys, $curr_string)) {
  $i = $i + 1;
  continue;
}
  $last_match_id = $lexicon[$curr_string];
  $result = $result . $last_match_id;
  $lexicon[$curr_string] = $last_match_id . '0';
  if (is_power_of_two($index)) {
  $new_lex = [];
  $new_keys = [];
  $j = 0;
  while ($j < count($keys)) {
  $curr_key = $keys[$j];
  $new_lex['0' . $curr_key] = $lexicon[$curr_key];
  $new_keys = _append($new_keys, '0' . $curr_key);
  $j = $j + 1;
};
  $lexicon = $new_lex;
  $keys = $new_keys;
}
  $new_key = bin_string($index);
  $lexicon[$new_key] = $last_match_id . '1';
  $keys = _append($keys, $new_key);
  $index = $index + 1;
  $curr_string = '';
  $i = $i + 1;
};
  return $result;
}
$sample = '1011001';
$decompressed = decompress_data($sample);
echo rtrim($decompressed), PHP_EOL;
